### PR TITLE
Store encoded key_pacakge

### PR DESCRIPTION
**Pull request by franziskuskiefer**
Encoding of key packages is a performance bottleneck.
In order to speed up things like tree hashes this PR
serializes the key package on creation and every update
and stores the serialized value in the key package itself.

### CODE REVIEW

File: src/key_packages/codec.rs [hunk 1]
```diff
@@ -16,14 +16,16 @@ impl Codec for KeyPackage {
...
+        let mut kp = KeyPackage {
             protocol_version,
             ciphersuite: Config::ciphersuite(cipher_suite_name)?,
             hpke_init_key,
             credential,
             extensions,
             signature,
+            encoded: Vec::new(),
         };
+        kp.encoded = kp.unsigned_payload()?;
```
Comment (kkohbrok) [inline src/key_packages/codec.rs:28]: It seems a bit odd to decode a key package and then immediately encode it again. But I assume this is a quirk our decoding mechanism?

**Review by kkohbrok — COMMENTED** [review]
Nice optimization for both signatures and hashes! One question, otherwise it looks good to me.

File: src/key_packages/codec.rs (hunk 1 above)
Comment (franziskuskiefer) [reply src/key_packages/codec.rs:28]: I didn't want to touch the `Cursor` to copy the encoded payload... There will be more versions of this in future that make all this nicer...

Comment (kkohbrok) [reply src/key_packages/codec.rs:28]: I took the liberty of creating issue #393. Looks good to me otherwise!

**Review by kkohbrok — APPROVED** [review]

**Pull request merged**